### PR TITLE
Support Data source permissions

### PR DESCRIPTION
**Pull request by peterholmberg**

### CODE REVIEW

File: pkg/api/datasources.go [hunk 1]
```diff
@@ -17,11 +17,27 @@ func GetDataSources(c *m.ReqContext) Response {
...
+	var datasources []*m.DataSource
+	if err := bus.Dispatch(&dsFilterQuery); err != nil {
+		if err != bus.ErrHandlerNotFound {
+			return Error(500, "Could not get datasources", err)
+		}
+
+		datasources = query.Result
+	} else {
+		datasources = dsFilterQuery.Result
+	}
```
Comment (torkelo) [inline pkg/api/datasources.go]: I don't think we need to filter here in the /api/datasources as this api is protected by org admin permissions. We do need this  in a next phase, but then I think we should break be above lines out and reuse it in getFrontendSettings

File: pkg/models/datasource.go [hunk 2]
```diff
@@ -187,6 +189,37 @@ type GetDataSourceByNameQuery struct {
...
-// EVENTS
-type DataSourceCreatedEvent struct {
+//  Permissions
+// ---------------------
+
+type DsPermissionType int
+
+const (
+	DsPermissionQuery DsPermissionType = 1 << iota
+	DsPermissionNoAccess
```
Comment (torkelo) [inline pkg/models/datasource.go]: Wonder if we should change place here and make DsPermissionNoAccess = 0

**Review by torkelo — COMMENTED** [review]
Looks good just some minor things

File: pkg/models/datasource.go [hunk 3]
```diff
@@ -187,6 +189,37 @@ type GetDataSourceByNameQuery struct {
...
+
+func (p DsPermissionType) String() string {
+	names := map[int]string{
+		int(DsPermissionQuery):    "Query",
+		int(DsPermissionNoAccess): "No Access",
+	}
+	return names[int(p)]
+}
+
+type HasRequiredDataSourcePermissionQuery struct {
```
Comment (torkelo) [inline pkg/models/datasource.go]: remove this, not used

**Comment by RestfulBlue** [comment]
guys am i properly understand what this pull request added possibility to set permissions for datasources and now it will be possible to restrict access to certain datasources for editors?

**Comment by RestfulBlue** [comment]
but why? it looks like it should be core functionality, as discussed in #2132 there currently no way to restrict access for editors and in real world we can let editor access only to a narrow circle of people, who can view all the data. Looks like in future versions also coming explore ui but without restricting access it's almost useless :( 

i understand what u need to make  some features for enterprise version to make enterprise version much popular and it is ok, but making data source permissions only for enterprise it is like making explore ui only for enterpise version, and every functionality where users can write own queries only for enterprise version

is there anything coming to somehow control access to datasource in opensource edition? for example pass some additional data using datasource implementation and handling it on database or set particular datasource for folder? so in editor inside particular folder can be used only specified datasource

**Pull request merged**